### ISSUE TITLE
Question: How to update a chart dynamically...?

### ISSUE THREAD

**Issue by markhm** [appreciated/apexcharts-flow#33]
Hi. I am trying to update the example ScatterChart dynamically. When calling `updateSeries(..)` on the chart object, I see updates arriving in the browser via Chrome's Console. 

```
    Received xhr message: ...
    Handling message from server
    Handling dependencies
    StateTree after applying changes: ...
```

This all looks good. 

But the actual chart does not update. Is there anything else I need to do...?

Thanks, Mark

**Comment by appreciated** [comment]
closed by bc0d8a02890e54a5c06b4fa9250bc8381f97ad9b

**Comment by appreciated** [comment]
To see a working example [check this out](https://github.com/appreciated/apexcharts-flow/blob/master/src/test/java/com/github/appreciated/apexcharts/examples/update/UpdateExample.java).

**Comment by markhm** [comment]
Hi appreciated,

Thank you for your response, also for all the other new things you've added the last week. 👍🏻

I confirm that the UpdateExample works when I run it based on your repository.

The strange thing is that I am seeing inconsistent/instable behaviour when integrating your UpdateExample in my application. 

My application follows a Simple App structure, as downloaded from https://vaadin.com/start/latest. I use Spring Boot 2.2.0-RELEASE and Vaadin 14.0.12, ApexCharts-Flow-2.0.0.beta7 with Java 8 on macOS. (I believe Spring Boot 2.2.x is not officially supported yet for Vaadin 14.0.y, but the previous release for Spring Boot did not give any better results.)

The MainLayout contains @Push. I have various individual pages with VerticalLayouts. 

Some of the strange behaviour I am seeing:
- Your UpdateExample does not work within my application.
- One other push example I've built myself works as expected on Safari, but requires a manual trigger (see the 'Show results' button) on Chome/Brave to show the line chart. 
- Yet another example using the exact same push mechanism (as follows), does nót require extra triggering on Chrome/Brave. 

```
...
        userInterface.access(() ->
        {
            addChart();
            userInterface.push();
        });
...
```

Do you have any idea what this could be or how to debug these kind of problems? 

No problem if not, I will post to the Vaadin Forum as well. Again, thanks for your effort. 

Best regards, Mark

<img width="975" alt="Screenshot 2019-11-04 at 22 15 38" src="https://user-images.githubusercontent.com/1430451/68159200-9f351200-ff51-11e9-8f1d-4552da0fa05f.png">

**Comment by appreciated** [comment]
Hi @markhm you're welcome 🙂! Try using different transport mechanisms than `Transport.WEBSOCKET_XHR` f.e. `@Push(transport = Transport.LONG_POLLING)`.

**Comment by markhm** [comment]
Hi appreciated,

Thanks again for your thoughts. 

Unfortunately, the problem persists. It seems, once a VerticalLayout is used via a MainLayout - so the @Push annotation is only referenced indirectly - the chart does not update. 

I've recreated the issue in my example repo, so you can easily reproduce, see:

```
    https://github.com/markhm/vaadin_examples
    mvn spring-boot:run
    http://localhost:8081/app
```

Regarding Atmosphere, I don't see any strange entries in the log section below, but I must confess the framework is new to me. What should I look for?

Best, Mark

```
20:34:37.849 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereHandler com.vaadin.flow.server.communication.PushAtmosphereHandler mapped to context-path: /*
20:34:37.850 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed the following AtmosphereInterceptor mapped to AtmosphereHandler com.vaadin.flow.server.communication.PushAtmosphereHandler
20:34:37.870 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere is using org.atmosphere.util.VoidAnnotationProcessor for processing annotation
20:34:37.873 INFO [restartedMain] org.atmosphere.util.ForkJoinPool: Using ForkJoinPool  java.util.concurrent.ForkJoinPool. Set the org.atmosphere.cpr.broadcaster.maxAsyncWriteThreads to -1 to fully use its power.
20:34:37.876 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed WebSocketProtocol org.atmosphere.websocket.protocol.SimpleHttpProtocol 
20:34:37.882 INFO [restartedMain] org.atmosphere.container.JSR356AsyncSupport: JSR 356 Mapping path /app
20:34:37.892 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installing Default AtmosphereInterceptors
20:34:37.893 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.CorsInterceptor : CORS Interceptor Support
20:34:37.893 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.CacheHeadersInterceptor : Default Response's Headers Interceptor
20:34:37.893 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.PaddingAtmosphereInterceptor : Browser Padding Interceptor Support
20:34:37.894 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.AndroidAtmosphereInterceptor : Android Interceptor Support
20:34:37.894 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Dropping Interceptor org.atmosphere.interceptor.HeartbeatInterceptor
20:34:37.894 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.SSEAtmosphereInterceptor : SSE Interceptor Support
20:34:37.894 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.JSONPAtmosphereInterceptor : JSONP Interceptor Support
20:34:37.895 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.JavaScriptProtocol : Atmosphere JavaScript Protocol
20:34:37.895 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.WebSocketMessageSuspendInterceptor : org.atmosphere.interceptor.WebSocketMessageSuspendInterceptor
20:34:37.895 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.OnDisconnectInterceptor : Browser disconnection detection
20:34:37.896 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.IdleResourceInterceptor : org.atmosphere.interceptor.IdleResourceInterceptor
20:34:37.896 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Set org.atmosphere.cpr.AtmosphereInterceptor.disableDefaults to disable them.
20:34:37.896 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor CORS Interceptor Support with priority FIRST_BEFORE_DEFAULT 
20:34:37.899 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Default Response's Headers Interceptor with priority AFTER_DEFAULT 
20:34:37.900 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Browser Padding Interceptor Support with priority AFTER_DEFAULT 
20:34:37.900 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Android Interceptor Support with priority AFTER_DEFAULT 
20:34:37.900 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor SSE Interceptor Support with priority AFTER_DEFAULT 
20:34:37.900 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor JSONP Interceptor Support with priority AFTER_DEFAULT 
20:34:37.900 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Atmosphere JavaScript Protocol with priority AFTER_DEFAULT 
20:34:37.901 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor org.atmosphere.interceptor.WebSocketMessageSuspendInterceptor with priority AFTER_DEFAULT 
20:34:37.901 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Browser disconnection detection with priority AFTER_DEFAULT 
20:34:37.901 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor org.atmosphere.interceptor.IdleResourceInterceptor with priority BEFORE_DEFAULT 
20:34:37.901 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using EndpointMapper class org.atmosphere.util.DefaultEndpointMapper
20:34:37.901 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using BroadcasterCache: org.atmosphere.cache.UUIDBroadcasterCache
20:34:37.901 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Default Broadcaster Class: org.atmosphere.cpr.DefaultBroadcaster
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Broadcaster Shared List Resources: false
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Broadcaster Polling Wait Time 100
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Shared ExecutorService supported: true
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Messaging ExecutorService Pool Size unavailable - Not instance of ThreadPoolExecutor
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Async I/O Thread Pool Size: 200
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using BroadcasterFactory: org.atmosphere.cpr.DefaultBroadcasterFactory
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using AtmosphereResurceFactory: org.atmosphere.cpr.DefaultAtmosphereResourceFactory
20:34:37.902 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using WebSocketProcessor: org.atmosphere.websocket.DefaultWebSocketProcessor
20:34:37.904 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Invoke AtmosphereInterceptor on WebSocket message true
20:34:37.904 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: HttpSession supported: true
20:34:37.904 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere is using DefaultAtmosphereObjectFactory for dependency injection and object creation
20:34:37.904 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere is using async support: org.atmosphere.container.JSR356AsyncSupport running under container: Apache Tomcat/9.0.27 using javax.servlet/3.0 and jsr356/WebSocket API
20:34:37.904 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere Framework 2.4.30.slf4jvaadin1 started.
20:34:37.908 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor  Track Message Size Interceptor using | with priority BEFORE_DEFAULT 
20:34:37.908 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereHandler com.vaadin.flow.server.communication.PushAtmosphereHandler mapped to context-path: /*
20:34:37.909 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed the following AtmosphereInterceptor mapped to AtmosphereHandler com.vaadin.flow.server.communication.PushAtmosphereHandler
20:34:37.911 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere is using org.atmosphere.util.VoidAnnotationProcessor for processing annotation
20:34:37.913 INFO [restartedMain] org.atmosphere.util.ForkJoinPool: Using ForkJoinPool  java.util.concurrent.ForkJoinPool. Set the org.atmosphere.cpr.broadcaster.maxAsyncWriteThreads to -1 to fully use its power.
20:34:37.913 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed WebSocketProtocol org.atmosphere.websocket.protocol.SimpleHttpProtocol 
20:34:37.914 INFO [restartedMain] org.atmosphere.container.JSR356AsyncSupport: JSR 356 Mapping path /frontend
20:34:37.914 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installing Default AtmosphereInterceptors
20:34:37.914 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.CorsInterceptor : CORS Interceptor Support
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.CacheHeadersInterceptor : Default Response's Headers Interceptor
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.PaddingAtmosphereInterceptor : Browser Padding Interceptor Support
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.AndroidAtmosphereInterceptor : Android Interceptor Support
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Dropping Interceptor org.atmosphere.interceptor.HeartbeatInterceptor
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.SSEAtmosphereInterceptor : SSE Interceptor Support
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.JSONPAtmosphereInterceptor : JSONP Interceptor Support
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.JavaScriptProtocol : Atmosphere JavaScript Protocol
20:34:37.915 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.WebSocketMessageSuspendInterceptor : org.atmosphere.interceptor.WebSocketMessageSuspendInterceptor
20:34:37.916 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.OnDisconnectInterceptor : Browser disconnection detection
20:34:37.916 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: 	org.atmosphere.interceptor.IdleResourceInterceptor : org.atmosphere.interceptor.IdleResourceInterceptor
20:34:37.916 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Set org.atmosphere.cpr.AtmosphereInterceptor.disableDefaults to disable them.
20:34:37.916 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor CORS Interceptor Support with priority FIRST_BEFORE_DEFAULT 
20:34:37.916 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Default Response's Headers Interceptor with priority AFTER_DEFAULT 
20:34:37.916 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Browser Padding Interceptor Support with priority AFTER_DEFAULT 
20:34:37.916 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Android Interceptor Support with priority AFTER_DEFAULT 
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor SSE Interceptor Support with priority AFTER_DEFAULT 
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor JSONP Interceptor Support with priority AFTER_DEFAULT 
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Atmosphere JavaScript Protocol with priority AFTER_DEFAULT 
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor org.atmosphere.interceptor.WebSocketMessageSuspendInterceptor with priority AFTER_DEFAULT 
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor Browser disconnection detection with priority AFTER_DEFAULT 
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor org.atmosphere.interceptor.IdleResourceInterceptor with priority BEFORE_DEFAULT 
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using EndpointMapper class org.atmosphere.util.DefaultEndpointMapper
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using BroadcasterCache: org.atmosphere.cache.UUIDBroadcasterCache
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Default Broadcaster Class: org.atmosphere.cpr.DefaultBroadcaster
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Broadcaster Shared List Resources: false
20:34:37.917 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Broadcaster Polling Wait Time 100
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Shared ExecutorService supported: true
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Messaging ExecutorService Pool Size unavailable - Not instance of ThreadPoolExecutor
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Async I/O Thread Pool Size: 200
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using BroadcasterFactory: org.atmosphere.cpr.DefaultBroadcasterFactory
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using AtmosphereResurceFactory: org.atmosphere.cpr.DefaultAtmosphereResourceFactory
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Using WebSocketProcessor: org.atmosphere.websocket.DefaultWebSocketProcessor
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Invoke AtmosphereInterceptor on WebSocket message true
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: HttpSession supported: true
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere is using DefaultAtmosphereObjectFactory for dependency injection and object creation
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere is using async support: org.atmosphere.container.JSR356AsyncSupport running under container: Apache Tomcat/9.0.27 using javax.servlet/3.0 and jsr356/WebSocket API
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Atmosphere Framework 2.4.30.slf4jvaadin1 started.
20:34:37.918 INFO [restartedMain] org.atmosphere.cpr.AtmosphereFramework: Installed AtmosphereInterceptor  Track Message Size Interceptor using | with priority BEFORE_DEFAULT 
20:34:37.936 INFO [restartedMain] org.apache.juli.logging.DirectJDKLog: Starting ProtocolHandler ["http-nio-8081"]
```

**Comment by appreciated** [comment]
@markhm when cloning your project, it does not seem to run out of the box. 

I just downloaded a fresh starter from Vaadin and the update seems to run without a problem. 
[my-starter-project.zip](https://github.com/appreciated/apexcharts-flow/files/3822197/my-starter-project.zip)

So there is probably something wrong with your (project-) setup 🤔.

**Comment by markhm** [comment]
Thank you for wanting to take a look. 

I've downloaded your starter, but that did not contain the chart update example. 

> So there is probably something wrong with your (project-) setup. 

Actually, here we see a manifestation of the stability problem [you've helped me with](https://github.com/vaadin/flow/issues/6036#issuecomment-536986765) before (!).

If you do the following, it will run and you will be able to see the problem. 

```
    // first clean thoroughly (in case you still have the repo)
    rm -Rf node_modules/
    rm package-lock.json
    mvn clean
    
    // build in full (missing step)
    mvn install

    // run
    mvn spring-boot:run

    http://localhost:8081/app
```

I've just tested with a clean version of https://github.com/markhm/vaadin_examples.git and this works.

**Comment by appreciated** [comment]
> I've downloaded your starter, but that did not contain the chart update 

you can find it under `example.com.packagename.myapp.spring.MainView`

**Comment by markhm** [comment]
This is indeed the example that I had already seen working. 

I had a hypothesis what might have been wrong, but I just had to reject that. Your example also works when the view is referred to from a MainLayout. 

Your last remark turns out to be spot on. The duplicate addon sources (a leftover from when I experimented directly with the  apexcharts-flow source code) turned out to be the problem. After I removed them, the chart update now works correctly. 

Much appreciated...! 🙏🏻
